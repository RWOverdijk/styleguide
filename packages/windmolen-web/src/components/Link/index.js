// @flow
import React from 'react';
import type ButtonTheme from '../Button';
import Base from '../Base';
import { getStyledButton, StyledIcon, ButtonText as StyledSpan } from '../Button/styles';
import { colors } from '../../globals';

type LinkProps = {
  asButton?: boolean,

  /** Buttom theme, c.f. see <Button /> */
  theme?: ButtonTheme,
  hideArrow?: boolean,
  children?: Node,
  icon?: string,
  small?: boolean
};

const StyledLink = Base.withComponent('a').extend`
  color: ${colors.warmGray};
  background-color: ${colors.transparent};
  text-decoration: underline;

  &:hover {
    color: ${colors.silver};
    background-color: ${colors.transparent};
    text-decoration: underline;
  }
`;

const StyledLinkAsButton = getStyledButton('a').extend`
  display: inline-block;

  &:hover {
    text-decoration: none;
  }
`;

const Link = (props: LinkProps) => props.asButton
  ? (
    <StyledLinkAsButton {...props} theme={props.theme}>
      {props.icon && <StyledIcon name={props.icon} />}

      <StyledSpan hideArrow={props.hideArrow}>{props.children}</StyledSpan>

      {!props.hideArrow && !props.icon && (
        <StyledIcon hide={props.hideArrow} name="arrow-right" />
      )}
    </StyledLinkAsButton>
  ) : <StyledLink fontSize="link" {...props} />;

Link.defaultProps = {
  theme: 'primary',
};

export default Link;
