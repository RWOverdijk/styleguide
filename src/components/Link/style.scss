/** @define Link */

@import "../../css/_global";
@import "../../css/mixins";

.Van-Link {
  @include color-fill($color-primary-bright);

  text-decoration: none;
  cursor: pointer;

  &:hover {
    @include color-fill($color-pink-primary);
  }

  &--secondary {
    @include color-fill($color-green-primary);

    &:hover {
      @include color-fill($color-green-primary);

      text-decoration: underline;
    }
  }
}
