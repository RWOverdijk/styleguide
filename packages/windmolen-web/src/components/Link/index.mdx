---
name: Link
menu: Elements
route: /Elements/Link
---

import { Playground, PropsTable, Link as DoczLink } from 'docz';
import Wrapper from '../Wrapper';
import Link from './';
import Paragraph from '../Paragraph';

## Props
See <DoczLink to="/web/Base/Pressable">Pressable</DoczLink>

## Basic usage
<Playground wrapper={Wrapper}>
  <Link href="#basic-usage">A link</Link>
</Playground>

## As button
<Playground wrapper={Wrapper}>
  <Link href="#as-button" variant="button-primary">A button-looking link</Link>
</Playground>

## As alternate button
<Playground wrapper={Wrapper}>
  <Link href="#as-alternate-button" variant="button-alternate">Different button-looking link</Link>
</Playground>

## As small button
<Playground wrapper={Wrapper}>
  <Link href="#as-small-button" variant="button-primary" small>Small button</Link>
</Playground>

## As button with icon
<Playground wrapper={Wrapper}>
  <Link href="#as-button-with-icon" variant="button-primary" hideIcon>Without icon</Link>
  <Link href="#as-button-with-icon" variant="button-primary" icon="windmill" iconPlacement="start">With icon on the left</Link>
  <Link href="#as-button-with-icon" variant="button-primary" icon="windmill">With icon on the right</Link>
</Playground>

## With custom component
<Playground wrapper={Wrapper}>
  <Link href="#with-custom-component" as={Paragraph}>With custom component</Link>
</Playground>
