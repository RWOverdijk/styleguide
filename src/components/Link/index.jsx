import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import './style.css';

class Link extends PureComponent {
  getClass(className) {
    let classes = ['Van-Link'];

    if (this.props.className) {
      classes.push(this.props.className);
    }

    if (this.props.type === 'secondary') {
      classes.push('Van-Link--secondary');
    }

    classes.push(className);

    return classes.join(' ').trim();
  }

  render() {
    const {
      className,
      ...props
    } = this.props;

    return <a className={this.getClass()} {...props}>{this.props.children}</a>;
  }
}

Link.propTypes = {
  className: PropTypes.string,
  type: PropTypes.string
};

export default Link;
