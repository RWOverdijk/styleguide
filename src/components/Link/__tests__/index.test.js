import React from 'react';
import renderer from 'react-test-renderer';
import Link from '../';

describe("<Link>", () => {
  it("renders primary Link", () => {
    const tree = renderer.create(
      <Link>My Link</Link>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders secondary Link component", () => {
    const tree = renderer.create(
      <Link type="secondary">My Link</Link>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with specific className", () => {
    const tree = renderer.create(
      <Link className="my-class">My Link</Link>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with specific attributes", () => {
    const tree = renderer.create(
      <Link href="#" title="My link" data-id="0">My Link</Link>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
