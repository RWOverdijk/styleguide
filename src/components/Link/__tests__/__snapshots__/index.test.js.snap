// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link> renders primary Link 1`] = `
<a
  className="Van-Link"
>
  My Link
</a>
`;

exports[`<Link> renders secondary Link component 1`] = `
<a
  className="Van-Link Van-Link--secondary"
>
  My Link
</a>
`;

exports[`<Link> renders with specific attributes 1`] = `
<a
  className="Van-Link"
  data-id="0"
  href="#"
  title="My link"
>
  My Link
</a>
`;

exports[`<Link> renders with specific className 1`] = `
<a
  className="Van-Link my-class"
>
  My Link
</a>
`;
